runtimes/core/src/api/gateway/websocket.rs

```diff
@@ -0,0 +1,329 @@
+use std::{collections::HashMap, str::FromStr};
+
+use anyhow::anyhow;
+use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine};
+use http::{header::SEC_WEBSOCKET_PROTOCOL, HeaderName, HeaderValue};
+use pingora::http::RequestHeader;
+use serde::{
+    de::{self, Visitor},
+    Deserialize,
+};
+
+const ENCORE_DEV_HEADERS: &str = "encore.dev.headers.";
+
+// hack to be able to have browsers send request headers when setting up a websocket
+// inspired by https://github.com/kubernetes/kubernetes/commit/714f97d7baf4975ad3aa47735a868a81a984d1f0
+pub fn update_headers_from_websocket_protocol(
+    upstream_request: &mut RequestHeader,
+) -> anyhow::Result<()> {
+    let headers = upstream_request
+        .headers
+        .get_all(SEC_WEBSOCKET_PROTOCOL)
+        .into_iter()
+        .cloned()
```

Comment:
Do we need to clone this into a vec? Or can we just iterate?

Reply:
We can't unfortunately have an iterator only here, as we then take a reference the headers map, and then later when we mutate the headers pingora takes a mutable reference.

I think this could be solved differently if pingora exposed a way to access the entry api to mutate in place, maybe. But since this only happens if the request is an upgrade request, and since 99.9999% of the time it will only be one header (e.g javascript/nodejs will send one comma separated header), it wont be too bad I think. I only added the "get_all" here in case several headers are sent instead of one comma separated header, which is supported according to the spec (they both mean the same thing), but ofc something we can also decide not to support and just parse one header.